src/unix/netmap-udp.c

```diff
@@ -0,0 +1,735 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "internal.h"
+
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#if defined(__MVS__)
+#include <xti.h>
+#endif
+#include <sys/un.h>
+
+#define NETMAP_WITH_LIBS
+#include <net/netmap_user.h>
+#include <net/ethernet.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+
+
+static uint16_t ip_checksum(struct ip* ip);
+static void uv__udp_netmap_recv_packet(uv_loop_t* loop, struct netmap_slot* slot, uint8_t* p);
+static size_t uv__udp_netmap_generate_udp(uv_loop_t* loop, unsigned int src_port, struct msghdr* h, uint8_t* pkt);
+static int uv__udp_netmap_send_udp(uv_loop_t* loop, uv_udp_send_t* req, struct netmap_ring* ring);
+static void uv__udp_netmap_run_completed(uv_loop_t* loop);
+static void uv__udp_netmap_io(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
+
+static const size_t ETH_LEN = sizeof(struct ether_header);
+static const size_t ETH_IP_LEN = ETH_LEN + sizeof(struct ip);
+static const size_t ETH_IP_UDP_LEN = ETH_IP_LEN + sizeof(struct udphdr);
+
+
+static uint16_t ip_checksum(struct ip* ip) {
+  int sum = 0;
+  size_t len = sizeof(struct ip);
+  int i;
+  size_t len_2 = len/2;
+
+  for (i = 0; i < len_2; i++) {
+    sum += *((uint16_t*)(ip) + i);
+  }
+
+  if (len_2 * 2 != len) {
+    sum += *((uint8_t*)(ip) + (len - 1));
+  }
+
+  while (sum > USHRT_MAX) {
+    sum -= USHRT_MAX;
+  }
+
+  return USHRT_MAX - sum;
+}
+
+static uint16_t udp_checksum(struct ip* ip, struct udphdr* udp, uint8_t* payload) {
+  int payload_len;
+  int sum;
+  int i;
+
+  sum = 0;
+  sum += *((uint16_t*)(&ip->ip_src.s_addr));
+  sum += *((uint16_t*)(&ip->ip_src.s_addr) + 1);
+  sum += *((uint16_t*)(&ip->ip_dst.s_addr));
+  sum += *((uint16_t*)(&ip->ip_dst.s_addr) + 1);
+  sum += htons(ip->ip_p);
+  sum += udp->len;
+
+  sum += udp->source;
+  sum += udp->dest;
+  sum += udp->len;
+
+  payload_len = ntohs(udp->len) - sizeof(*udp);
+  for (i = 0; i < payload_len / 2; ++i) {
+    sum += *((uint16_t*)(payload) + i);
+  }
+
+  if (payload_len % 2 == 1) {
+    sum += *((uint8_t*)(payload) + (payload_len - 1));
+  }
+
+  while (sum > USHRT_MAX) {
+    sum -= USHRT_MAX;
+  }
+
+  return USHRT_MAX - sum;
+}
+
+
+static void uv__udp_netmap_recv_packet(uv_loop_t* loop, struct netmap_slot* slot, uint8_t* p) {
+  struct ether_header* eth;
+  struct ip* ip;
+  struct udphdr* udp;
+  uint8_t* payload;
+
+  if (slot->len < ETH_IP_UDP_LEN) {
+    return;
+  }
+
+  eth = (struct ether_header*)p;
+
+  if (eth->ether_type != htons(ETHERTYPE_IP)) {
+    return;
+  }
+
+  ip = (struct ip*)(p + ETH_LEN);
+
+  if (ip->ip_p != IPPROTO_UDP) {
+    return;
+  }
+
+  udp = (struct udphdr*)(p + ETH_IP_LEN);
+
+  // todo: verify checksum?
```

Comment:
Seems like this would be an easy one to knock out given you've implemented `udp_checksum` above?